src/object/obj_tx.c

```diff
@@ -640,154 +673,774 @@ dc_tx_open(tse_task_t *task)
 }
 
 struct tx_commit_cb_args {
-	struct dc_tx	*tcca_tx;
-	crt_rpc_t	*tcca_req;
+	struct dc_tx		*tcca_tx;
+	crt_rpc_t		*tcca_req;
+	daos_tx_commit_t	*tcca_args;
 };
 
 static int
 dc_tx_commit_cb(tse_task_t *task, void *data)
 {
-	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct tx_commit_cb_args *tcca = data;
 	struct dc_tx		 *tx = tcca->tcca_tx;
 	crt_rpc_t		 *req = tcca->tcca_req;
+	struct obj_cpd_out	 *oco = crt_reply_get(req);
+	uint64_t		 *sub_epochs;
 	int			  rc = task->dt_result;
+	bool			  locked = true;
 
 	D_MUTEX_LOCK(&tx->tx_lock);
 
+	if (rc == 0) {
+		rc = oco->oco_ret;
+		if (rc == 0) {
+			int	*sub_rets = oco->oco_sub_rets.ca_arrays;
+
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			rc = sub_rets[0];
+		}
+	}
+
 	if (rc == 0) {
 		tx->tx_status = TX_COMMITTED;
+		tx->tx_retry = 0;
+		dc_tx_cleanup(tx);
+
+		sub_epochs = oco->oco_sub_epochs.ca_arrays;
+		if (tx->tx_epoch.oe_value == 0)
+			/* FIXME: currently, we pack one DTX in one CPD RPC. */
+			tx->tx_epoch.oe_value = sub_epochs[0];
+
+		if (tx->tx_epoch.oe_value == 0)
+			D_WARN("Server forgot to reply epoch for TX "DF_DTI"\n",
+			       DP_DTI(&tx->tx_id));
+
 		goto out;
 	}
 
 	tx->tx_status = TX_FAILED;
 
-	/* Required to restart the TX with newer epoch. */
-	if (rc == -DER_TX_RESTART)
+	/* Be required to restart the TX with newer epoch. */
+	if (rc == -DER_TX_RESTART) {
+		tx->tx_pm_ver = oco->oco_map_version;
+		task->dt_result = rc;
+		/* Set retry flag even if it is restarted. */
+		tx->tx_retry = 1;
```

Comment:
Could you elaborate this, please? Don't we need to call prepare in this case, considering the objects/keys being read/written may differ in the new TX generation?

Reply:
Will fix it in new commit